kubelet: add setting for configuring cloudProvider

jpculp opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
https://github.com/bottlerocket-os/bottlerocket/issues/1459


**Description of changes:**

Adds a new setting `kubernetes.cloud-provider` for configuring whether the cloud provider is `aws` or `external`.
Prior to this, the argument was hard-coded to `aws`.

Adds a migration for the new `kubernetes.cloud-provider` setting.

**Testing done:**
- [x] Build `aws-k8s-1.19` ami.
- [x] Launch and connect to EKS cluster.
- [x] Deploy something to the node.
- [x] Connect to admin container and `sudo sheltie`.
- [x] Verify that there are no kubelet errors.
- [x] Repeat the same steps for `aws-k8s-1.18`.
- [x] Repeat the same steps for `aws-k8s-1.17`.
- [x] Repeat the same steps for `aws-k8s-1.16`.
- [x] Verify the migration works as expected.
- [x] Verify variant without a default cloud provider specified defaults to external.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

webern commented on packages/kubernetes-1.16/kubelet-exec-start-conf:
```diff
@@ -2,14 +2,12 @@
-{{~else}}
-    --cloud-provider "" \
 {{~/unless}}
+    --cloud-provider "${CLOUD_PROVIDER}" \
```
Is there a reason to use an environment variable instead of templating the setting value itself?

jpculp replied:
I was following some existing code, but since it's not necessary I've removed the environment variable.

webern commented on packages/kubernetes-1.17/kubelet-env:
```diff
@@ -2,3 +2,4 @@ NODE_IP={{settings.kubernetes.node-ip}}
 NODE_LABELS={{join_map "=" "," "no-fail-if-missing" settings.kubernetes.node-labels}}
 NODE_TAINTS={{join_map "=" "," "no-fail-if-missing" settings.kubernetes.node-taints}}
 POD_INFRA_CONTAINER_IMAGE={{settings.kubernetes.pod-infra-container-image}}
+CLOUD_PROVIDER={{settings.kubernetes.cloud-provider}}
```
Same question as above. Curious whether this needs to be an environment variable whether its value should be written to the conf.

bcressey replied:
Everything but `POD_INFRA_CONTAINER_IMAGE` can be in the conf template now. The environment variable approach was all we had before ef41a6b1, but isn't as flexible.

We don't have to move the other settings here, but shouldn't add more.

jpculp replied:
Updated.

webern commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -723,3 +723,48 @@ mod test_kubernetes_quantity_value {
+/// kubelet.  It stores the original string and makes it accessible through standard traits.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct KubernetesCloudProvider {
+    inner: String,
```
Maybe an enum would be better. Not a strong preference but it looks like we have enums for other settings.

zmrow replied:
I think I'd reach for an enum as well. :)

jpculp replied:
I'm not crazy about the enum for something this small. This block of code used [KubernetesAuthenticationMode](https://github.com/bottlerocket-os/bottlerocket/blob/64e679747e3af8a029df7e66ae5dfc8b6c220e0c/sources/models/src/modeled_types/kubernetes.rs#L356) as a reference.

jpculp replied:
I'm open to replacing it with an enum later, but we would want to address https://github.com/bottlerocket-os/bottlerocket/issues/1501 first.

zmrow commented on README.md:
```diff
@@ -308,6 +308,7 @@ The following settings can be optionally set to customize the node labels and ta
 The following settings are optional and allow you to further configure your cluster.
 * `settings.kubernetes.cluster-domain`: The DNS domain for this cluster, allowing all Kubernetes-run containers to search this domain before the host's search domains.  Defaults to `cluster.local`.
 * `settings.kubernetes.standalone-mode`: Whether to run the kubelet in standalone mode, without connecting to an API server.  Defaults to `false`.
+* `settings.kubernetes.cloud-provider`: The cloud platform for this cluster. Defaults to `aws` for AWS variants, and `external` for other variants.
```
We should probably say "cloud provider" rather than "cloud platform".

jpculp replied:
I was trying to avoid repeating the name of the setting, but if folks feel strongly or have a less redundant suggestion I'm open to changing it.

etungsten replied:
I think we should keep it in line with the term that the k8s official doc uses to remove any ambiguity on what this is for. See https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/

jpculp commented:
- Removed the environment variable logic.

jpculp commented:
- Tested all other support kubernetes versions.
- Performed migration test with 1.19 instance.

bcressey commented on packages/kubernetes-1.18/kubelet-exec-start-conf:
```diff
@@ -2,14 +2,12 @@
-{{~else}}
-    --cloud-provider "" \
 {{~/unless}}
+    --cloud-provider {{settings.kubernetes.cloud-provider}} \
```
We need the old logic here so that we pass an empty string to cloud-provider if standalone mode is enabled.

jpculp committed:
k8s: add setting for configuring cloudProvider

Adds a new setting `kubernetes.cloud-provider` for configuring whether
the cloud provider is `aws` or `external`. Prior to this, the argument
was hard-coded to `aws`.

jpculp committed:
migrations: add migration for adding `kubernetes.cloud-provider`

Adds a migration for the new `kubernetes.cloud-provider` setting.

jpculp commented:
- Reverted back to unless+else logic in `kubelet-exec-start-conf`.
- Added default setting of `external` if `settings.kubernetes.cloud-provider` is absent from the variant's defaults.
- Replaced the word `platform` with `provider`.

zmrow reviewed (approved):
:safety_vest: 

Code looks good.  Before you merge - let's make sure to build an image without the `cloud-provider` setting and make sure that it defaults to "external" like we think it should.

jpculp commented:
Tested migration with a variant that did not have a default cloud provider setting set. Launched as `--cloud-provider aws` with 1.0.8 and after upgrading to a modified 1.1.0 variant launched as `--cloud-provider external`.
